pkg/sql/scanner/plpgsql_scan.go

```diff
@@ -0,0 +1,207 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package scanner
+
+import (
+	"unicode/utf8"
+	"unsafe"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/lexbase"
+	"github.com/cockroachdb/cockroach/pkg/sql/plpgsql/parser/plpgsqllexbase"
+)
+
+// PLPGSQLScanner is a scanner with a PLPGSQL specific scan function
+type PLPGSQLScanner struct {
+	Scanner
+}
+
+// Scan scans the next token and populates its information into lval.
+// This scan function contains rules for plpgsql.
+func (s *PLPGSQLScanner) Scan(lval ScanSymType) {
+	lval.SetID(0)
+	lval.SetPos(int32(s.pos))
+	lval.SetStr("EOF")
+
+	if _, ok := s.skipWhitespace(lval, true); !ok {
+		return
+	}
+
+	ch := s.next()
+	if ch == eof {
+		lval.SetPos(int32(s.pos))
+		return
+	}
+
+	lval.SetID(int32(ch))
+	lval.SetPos(int32(s.pos - 1))
+	lval.SetStr(s.in[lval.Pos():s.pos])
+
+	switch ch {
+	case '$':
+		if s.scanDollarQuotedString(lval) {
+			lval.SetID(plpgsqllexbase.SCONST)
+			return
+		}
+		return
+
+	case identQuote:
+		// "[^"]"
+		if s.scanString(lval, identQuote, false /* allowEscapes */, true /* requireUTF8 */) {
+			lval.SetID(plpgsqllexbase.IDENT)
+		}
+		return
+
+	case singleQuote:
+		// '[^']'
+		if s.scanString(lval, ch, false /* allowEscapes */, true /* requireUTF8 */) {
+			lval.SetID(plpgsqllexbase.SCONST)
+		}
+		return
+
+	case 'b':
+		// Bytes?
+		if s.peek() == singleQuote {
+			// b'[^']'
+			s.pos++
+			if s.scanString(lval, singleQuote, true /* allowEscapes */, false /* requireUTF8 */) {
+				lval.SetID(plpgsqllexbase.BCONST)
+			}
+			return
+		}
+		s.scanIdent(lval)
+		return
+
+	case '.':
+		switch t := s.peek(); {
+		case t == '.': // ..
+			s.pos++
+			lval.SetID(plpgsqllexbase.DOT_DOT)
+			return
+		case lexbase.IsDigit(t):
+			s.scanNumber(lval, ch)
+			return
+		}
+		return
+
+	case '!':
+		switch s.peek() {
+		case '=': // !=
+			s.pos++
+			lval.SetID(plpgsqllexbase.NOT_EQUALS)
+			return
+		}
+		return
+
+	case '<':
+		switch s.peek() {
+		case '<': // <<
+			s.pos++
+			lval.SetID(plpgsqllexbase.LESS_LESS)
+			return
+		case '=': // <=
+			s.pos++
+			lval.SetID(plpgsqllexbase.LESS_EQUALS)
+			return
+		}
+		return
+
+	case '>':
+		switch s.peek() {
+		case '>': // >>
+			s.pos++
+			lval.SetID(plpgsqllexbase.GREATER_GREATER)
+			return
+		case '=': // >=
+			s.pos++
+			lval.SetID(plpgsqllexbase.GREATER_EQUALS)
+			return
+		}
+		return
+
+	case ':':
+		switch s.peek() {
+		case ':':
+			s.pos++
+			lval.SetID(plpgsqllexbase.TYPECAST)
+			return
+		case '=':
+			s.pos++
+			lval.SetID(plpgsqllexbase.COLON_EQUALS)
+			return
+		}
+		return
+
+	default:
+		if lexbase.IsDigit(ch) {
+			s.scanNumber(lval, ch)
+			return
+		}
+		if lexbase.IsIdentStart(ch) {
+			s.scanIdent(lval)
+			return
+		}
+	}
+	// Everything else is a single character token which we already initialized
+	// lval for above.
+}
+
+func (s *PLPGSQLScanner) scanIdent(lval ScanSymType) {
+	s.pos--
```

Comment:
I think this method is almost the same as the first half of the one in `scanner.go`,  can we share something?